rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,223 @@
+
+## Detailed Design
+
+### Subgraph Definition
+
+Part of the power of fulltext search indexes are their flexibility, so 
+it is important to expose a simple interface to facilitate useful 
+applications of the index and ulimately aim to reduce the need to create 
+new subgraphs for the express purpose of updating fulltext search fields 
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+
+The proposed process of adding a fulltext search index to an Entity 
+involves adding a special field to that Entity in the GraphQL schema of 
+the subgraph's definition. The subgraph developer will create a new virtual 
+field of type String that is marked with the special `@FullTextSearch` directive. 
+The directive will have two required parameters for specifying the language
+ and the fields to be incorporated. Verification of the index definition 
+ will ensure that all fields referenced are valid String type fields on 
+ that entity. To be clear, the proposed interface will add the fulltext 
+ search definitions to the subgraph deployment hash, so adding or updating 
+ a fulltext search index will require an update to the subgraph.  
+ With subgraph composition it will be possible to easily create new subgraphs 
+ that add specific fulltext search indexes to an existing subgraph. 
+
+Example fulltext search field definition:
+```graphql
+type ExampleItem @entity {
+    id: ID!,
+    name: String!,
+    description: String!,
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @fullTextSearchIndex(language: "english", fields: [{name: "name", weight: 5}, {name: "description", weight: 3}, {name: "specs", weight: 1}])
+}
+```  
+
+### GraphQL Query interface
+
+For each fulltext index three search algorithms will be made available to 
+the end user: a simple text query, a standard ranking, and a cover 
+density ranking algorithm: 
+  - `query`: search for the string in the indexed document. Several 
+    operators available: and, or, and proximity (`&`, `|`, `<->`.) 
+    The proximity operator allows one to specify max distance: 
+    3 words apart → `<3>`. (filter suffix = ``)
+  - `standard ranking`: ranking based on the number of matching lexemes. 
+    (filter suffix = `_rank`)
+  - `cover density ranking`: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.
+    (filter suffix = `_rank_cd`)
+    
+Example Usage in a query (references the example Entity definition from above section): 
+```graphql
+query {
+    exampleItems(first: 10, where: {itemSearch_rank_cd: "sarcastic"}) {
```

Comment:
I dislike this API. It exposes a lot of complexity to the developer. The API should default to simple and ergonomic with more complex capabilities added as optional parameters.

Reply:
I've updated the API to use more general, self-describing language: `rank_cd` -> `proximity_rank`.  Keeping in mind the query can be made without any suffixes do you think this reduces the complexity you pointed out?

There is also the complexity of using operators in the filter: `&`, `|` and `<->`.  I think these operators are general and clear enough, but their usage would need to be conveyed through documentation or schema descriptions as it may not be obvious.